contracts/utils/TimestampedHashRegistry.sol

```diff
@@ -0,0 +1,116 @@
+import "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "./interfaces/ITimestampedHashRegistry.sol";
+
+contract TimestampedHashRegistry is
+    Ownable,
+    EIP712,
+    SelfMulticall,
+    ITimestampedHashRegistry
+{
+    using ECDSA for bytes32;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    mapping(bytes32 => EnumerableSet.AddressSet) private _hashTypeToSigners;
+
+    mapping(bytes32 => bytes32) public hashTypeToHash;
+
+    mapping(bytes32 => uint256) public hashTypeToTimestamp;
+
+    bytes32 private constant _SIGNED_HASH_TYPE_HASH =
+        keccak256(
+            "SignedHash(bytes32 hashType,bytes32 hash,uint256 timestamp)"
+        );
+
+    constructor() EIP712("TimestampedHashRegistry", "1.0.0") {}
+
+    function setupSigners(
+        bytes32 hashType,
+        address[] calldata signers
+    ) external override onlyOwner {
+        require(signers.length != 0, "Signers is empty");
+        require(
+            _hashTypeToSigners[hashType].length() == 0,
+            "Hash type signers is not empty"
+        );
+        for (uint256 ind = 0; ind < signers.length; ind++) {
+            _addSigner(hashType, signers[ind]);
+        }
+        emit SetupSigners(hashType, signers);
+    }
+
+    function _addSigner(bytes32 hashType, address signer) private {
+        require(hashType != bytes32(0), "Hash type is zero");
+        require(signer != address(0), "Signer is zero");
+        require(
+            _hashTypeToSigners[hashType].add(signer),
+            "Signer already exists"
+        );
+    }
+
+    function addSigner(
+        bytes32 hashType,
+        address signer
+    ) external override onlyOwner {
+        _addSigner(hashType, signer);
+        emit AddedSigner(hashType, signer);
+    }
+
+    function removeSigner(
+        bytes32 hashType,
+        address signer
+    ) external override onlyOwner {
+        require(hashType != bytes32(0), "Hash type is zero");
+        require(signer != address(0), "Signer is zero");
+        require(
+            _hashTypeToSigners[hashType].remove(signer),
+            "Signer does not exist"
+        );
+        emit RemovedSigner(hashType, signer);
+    }
+
+    function getSigners(
+        bytes32 hashType
+    ) external view override returns (address[] memory signers) {
+        signers = _hashTypeToSigners[hashType].values();
+    }
+
+    function registerHash(
+        bytes32 hashType,
+        bytes32 hash,
+        uint256 timestamp,
+        bytes[] calldata signatures
+    ) external override {
+        require(hashType != bytes32(0), "Hash type is zero");
+        EnumerableSet.AddressSet storage signers = _hashTypeToSigners[hashType];
+        require(signers.length() != 0, "Signers have not been set");
+        require(
+            signatures.length == signers.length(),
+            "Invalid number of signatures"
+        );
```

Comment:
```suggestion
        );
        require(timestamp > hashTypeToTimestamp[hashType], "Timestamp is not newer");
```

Reply:
Applying the same logic, should we also check that hash is != to hash on chain?

Reply:
Makes sense since I cant think of any case where we would need to reregister the same hash with a newer timestamp.

Reply:
I can think of one scenario. Say we have a valid signed hash that we have registered. Then we sign another hash with a more recent timestamp. Before registering it, we realize that the new hash is wrong and we want to keep the old one. We would want to update with the same hash with an even more recent timestamp to invalidate the wrong hash so that it can't be registered by anyone.